File: fdtd/src/matlab/linux/fdtd_first_version_01.m
%% Одномерный FDTD. Версия 1.0
clear

% Волновое сопротивление свободного пространства
W0 = 120 * pi;

% Число Куранта
Sc = 1.0;

% Время расчета в отсчетах
maxTime = 1000;

% Размер области моделирования в отсчетах
maxSize = 200;

% Положение датчика, регистрирующего поля
probePos = 50;

% Начальные условия
Ez = zeros (1, maxSize);
Hy = zeros (1, maxSize);

% Значение компонент поля в предыдущий момент времени
% Ez(q - 1), Hy(q - 1/2)
Ez_prev = zeros (size (Ez));
Hy_prev = zeros (size (Hy));

% Поле, зарегистрированное в датчике в зависимости от времени
probeTimeEz = zeros (1, maxTime);

for t = 1: maxTime
    Ez_prev = Ez;
    Hy_prev = Hy;
    
    % Расчет компоненты поля H
    for m = 1: maxSize - 1
        Hy(m) = Hy_prev(m) + (Ez_prev(m + 1) - Ez_prev(m)) * Sc / W0;
    end

    % Расчет компоненты поля E
    for m = 2: maxSize
        Ez(m) = Ez_prev(m) + (Hy(m) - Hy(m - 1)) * Sc * W0;
    end

    % Источник возбуждения
    Ez(1) = exp (-(t - 30.0) ^ 2 / 100.0);

    % Регистрация поля в точке
    probeTimeEz(t) = Ez(probePos);
end

figure
plot (probeTimeEz)
xlabel ('t, отсчет')
ylabel ('Ez, В/м')
grid on

File: src/python/fdtd_first_version_01.py
# coding: utf-8
'''
История изменений:
    * Первая версия программы с использованием метода FDTD.
'''

import numpy as np
import matplotlib.pyplot as plt


if __name__ == '__main__':
    # Волновое сопротивление свободного пространства
    W0 = 120.0 * np.pi

    # Число Куранта
    Sc = 1.0

    # Время расчета в отсчетах
    maxTime = 1000

    # Размер области моделирования в отсчетах
    maxSize = 200

    # Положение датчика, регистрирующего поле
    probePos = 50

    # Положение источника
    sourcePos = 75

    Ez = np.zeros(maxSize)
    Hy = np.zeros(maxSize)

    # Поле, зарегистрированное в датчике в зависимости от времени
    probeTimeEz = np.zeros(maxTime)
    probeTimeEz[0] = Ez[probePos]

    for q in range(1, maxTime):
        # Расчет компоненты поля H
        for m in range(0, maxSize - 1):
            Hy[m] = Hy[m] + (Ez[m + 1] - Ez[m]) * Sc / W0

        # Расчет компоненты поля E
        for m in range(1, maxSize):
            Ez[m] = Ez[m] + (Hy[m] - Hy[m - 1]) * Sc * W0

        # Источник возбуждения
        Ez[sourcePos] += np.exp(-(q - 0.5 - 30.0) ** 2 / 100.0)

        # Регистрация поля в точке
        probeTimeEz[q] = Ez[probePos]

    # Отображение сигнала, сохраненного в датчике
    time = np.arange(maxTime)
    fig, ax = plt.subplots()
    ax.set_xlim(0, maxTime)
    ax.set_ylim(-1.1, 1.1)
    ax.set_xlabel('q, отсчет')
    ax.set_ylabel('Ez, В/м')
    ax.plot(time, probeTimeEz)
    ax.grid()
    plt.show()
